packages/typespec-test/test/batch_modular/generated/typespec-ts/src/api/pagingHelpers.ts

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  getPagedAsyncIterator,
+  PagedAsyncIterableIterator as CorePagedAsyncIterableIterator,
+  PagedResult,
+  PageSettings as CorePageSettings,
+} from "@azure/core-paging";
+import {
+  Client,
+  createRestError,
+  PathUncheckedResponse,
+} from "@azure-rest/core-client";
+import {
+  PageSettings,
+  PagedAsyncIterableIterator,
+} from "../models/pagingTypes.js";
+
+export function buildPagedAsyncIterator<
+  TElement,
+  TResponse extends PathUncheckedResponse = PathUncheckedResponse
+>(
+  client: Client,
+  getInitialResponse: () => PromiseLike<TResponse>,
+  processResponseBody: (result: TResponse) => Promise<unknown>
+): PagedAsyncIterableIterator<TElement> {
+  let firstRun = true;
+  let itemName: string, nextLinkName: string | undefined;
+  const firstPageLinkPlaceholder = "";
+  const pagedResult: PagedResult<TElement[]> = {
+    firstPageLink: firstPageLinkPlaceholder,
+    getPage: async (pageLink: string) => {
+      const result =
+        firstRun && pageLink === firstPageLinkPlaceholder
+          ? await getInitialResponse()
+          : await client.pathUnchecked(pageLink).get();
+      if (firstRun) {
+        const pageInfo = getPaginationProperties(result);
+        itemName = pageInfo.itemName;
+        nextLinkName = pageInfo.nextLinkName;
+      }
+      firstRun = false;
+      checkPagingRequest(result);
+      const results = await processResponseBody(result as TResponse);
+      const nextLink = getNextLink(results, nextLinkName);
+      const values = getElements<TElement>(results, itemName);
+      return {
+        page: values,
+        nextPageLink: nextLink,
+      };
+    },
+    byPage: (settings?: PageSettings) => {
+      const { continuationToken } = settings ?? {};
+      return getPageAsyncIterator(pagedResult, {
+        pageLink: continuationToken,
+      }) as any;
+    },
+  };
+  const iter: CorePagedAsyncIterableIterator<TElement> =
+    getPagedAsyncIterator(pagedResult);
+
+  return {
+    next() {
+      return iter.next();
+    },
+    [Symbol.asyncIterator]() {
+      return this;
+    },
+    byPage: (settings?: PageSettings) => {
+      return iter.byPage(
+        settings as CorePageSettings
+      ) as unknown as AsyncIterableIterator<
+        TElement[] & { continuationToken?: string }
```

Comment:
I wonder if it would be better for us to standardize on the page always being an object to make it more coherent when there are other page-level properties (e.g. quota remaining and batch metadata)

Perhaps something like

```ts
{
  values: TElement[];
  continuationToken? string;
}
```

Alternatively, if we didn't want to pin the values to being an array, we could also consider making this page object itself be an iterator e.g.

```ts
Iterable<TElement> &
{
  continuationToken? string;
}
```

Reply:
Both definitions work for me, but I updated the change to the following because it would not introduce any breakings for HLC.

```ts
export type ContinuablePage<TElement, TPage = TElement[]> = TPage & {
  /**
   * The token that keeps track of where to continue the iterator
   */
  continuationToken?: string;
};
```